mobile_verifier/src/coverage.rs

```diff
@@ -0,0 +1,725 @@
+    poc_mobile::{self as proto, CoverageObjectValidity, SignalLevel as SignalLevelProto},
+};
+use mobile_config::client::AuthorizationClient;
+use retainer::Cache;
+use rust_decimal::Decimal;
+use rust_decimal_macros::dec;
+use sqlx::{FromRow, Pool, Postgres, Transaction, Type};
+use tokio::sync::mpsc::Receiver;
+use uuid::Uuid;
+
+#[derive(Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Type)]
+#[sqlx(type_name = "signal_level")]
+#[sqlx(rename_all = "lowercase")]
+pub enum SignalLevel {
+    None,
+    Low,
+    Medium,
+    High,
+}
+
+impl From<SignalLevelProto> for SignalLevel {
+    fn from(level: SignalLevelProto) -> Self {
+        match level {
+            SignalLevelProto::High => Self::High,
+            SignalLevelProto::Medium => Self::Medium,
+            SignalLevelProto::Low => Self::Low,
+            SignalLevelProto::None => Self::None,
+        }
+    }
+}
+
+pub struct CoverageDaemon {
+    pool: Pool<Postgres>,
+    auth_client: AuthorizationClient,
+    coverage_objs: Receiver<FileInfoStream<CoverageObjectIngestReport>>,
+    file_sink: FileSinkClient,
+}
+
+impl CoverageDaemon {
+    pub fn new(
+        pool: Pool<Postgres>,
+        auth_client: AuthorizationClient,
+        coverage_objs: Receiver<FileInfoStream<CoverageObjectIngestReport>>,
+        file_sink: FileSinkClient,
+    ) -> Self {
+        Self {
+            pool,
+            auth_client,
+            coverage_objs,
+            file_sink,
+        }
+    }
+
+    pub async fn run(mut self, shutdown: triggered::Listener) -> anyhow::Result<()> {
+        tokio::spawn(async move {
+            loop {
+                tokio::select! {
+                    _ = shutdown.clone() => {
+                        tracing::info!("CoverageDaemon shutting down");
+                        break;
+                    }
+                    Some(file) = self.coverage_objs.recv() => self.process_file(file).await?,
+                }
+            }
+
+            Ok(())
+        })
+        .map_err(anyhow::Error::from)
+        .and_then(|result| async move { result })
+        .await
+    }
+
+    async fn process_file(
+        &mut self,
+        file: FileInfoStream<CoverageObjectIngestReport>,
+    ) -> anyhow::Result<()> {
+        tracing::info!("Processing coverage object file {}", file.file_info.key);
+
+        let mut transaction = self.pool.begin().await?;
+        let reports = file.into_stream(&mut transaction).await?;
+
+        let mut validated_coverage_objects = pin!(CoverageObject::validate_coverage_objects(
+            &self.auth_client,
+            reports
+        ));
+
+        while let Some(coverage_object) = validated_coverage_objects.next().await.transpose()? {
+            coverage_object.write(&self.file_sink).await?;
+            coverage_object.save(&mut transaction).await?;
+        }
+
+        self.file_sink.commit().await?;
+        transaction.commit().await?;
+
+        Ok(())
+    }
+}
+
+pub struct CoverageObject {
```

Comment:
A CoverageObject is also defined in filestore ( without the validity field ).  Why repeat the whole thing here ? Instead why not something like:

pub struct CoverageObject {
object: Filestore::coverage:CoverageObject,
validity: CoverageObjectValidity,
}

I guess you have the second CoverageObject instance here to allow for easier `impl`s but duplicating their entire structure would not be an ideal pattern

I recently just noticed the same pattern in my speedtest PR and am looking to address that in a similar means to the above

Reply:
Good point. Addressed